### PR TITLE
PR for #251 - Convert JavaScript to Kotlin

### PR DESCRIPTION
**Pull request by chess-equality**
#251

For Traces:

Converted:
1. `displayTraces()`
2. `displayTraceStack()`

Pending:
1. `displayInnerTraces()` (to be merged)
2. `displaySpanInfo()`

Others:
1. Check `todo`s

### CODE REVIEW

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 1]
```diff
@@ -1,28 +1,91 @@
...
+    init {
+        console.log("Traces tab started")
+        setupUI()
+        eb.onopen = {
+            js("portalConnected()")
+            eb.registerHandler(DisplayTraces(portalUuid)) { error: String, message: dynamic ->
+                console.log("########## Calling displayTraces...")
+                val body: dynamic = message.body
+                val traceResult: TraceResult = Json.decodeFromDynamic(body)
+                console.log(">>>>>>>>>> artifactQualifiedName = ${traceResult.artifactQualifiedName}")
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: You have these ">>>>>" console loggings everywhere. Please remove them when you commit/create PRs. They should only be for you to debug.

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 2]
```diff
@@ -1,28 +1,91 @@
...
     private val portalUuid: String,
     private val traceOrderType: TraceOrderType
 ) {
+    init {
+        console.log("Traces tab started")
+        setupUI()
+        eb.onopen = {
+            js("portalConnected()")
+            eb.registerHandler(DisplayTraces(portalUuid)) { error: String, message: dynamic ->
+                console.log("########## Calling displayTraces...")
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: Same with these. If you are going to add logging that you want to remain it needs to be helfpul and formatted the same as other logging.

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 3]
```diff
@@ -52,4 +115,428 @@ class TracesPage(
...
+                "appUuid" to appUuid,
+                "artifactQualifiedName" to rootArtifactQualifiedName,
+                "traceId" to traceId,
+                "segmentId" to segmentId,
+                "spanId" to spanId
+            )
+        )
+    }
+
+    fun clickedBackToTraces() {
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: You need to implement this method for it to work. See: https://github.com/sourceplusplus/SourceMarker/blob/327e56aa516ceca9947ca25858735bfab52e8458/portal/frontend/src/main/kotlin/com/sourceplusplus/portal/template/Navbar.kt#L47

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 4]
```diff
@@ -52,4 +115,428 @@ class TracesPage(
...
+        eb.send(
+            ClickedDisplayTraces,
+            json(
+                "portalUuid" to portalUuid
+            )
+        )
+    }
+
+    fun clickedBackToTraceStack() {
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: You need to implement this method for it to work. See: https://github.com/sourceplusplus/SourceMarker/blob/327e56aa516ceca9947ca25858735bfab52e8458/portal/frontend/src/main/kotlin/com/sourceplusplus/portal/template/Navbar.kt#L55

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 5]
```diff
@@ -52,4 +115,428 @@ class TracesPage(
...
+        console.log(
+            """>>>>>>>>>> Displaying traces - Artifact: ${traceResult.artifactSimpleName} 
+            - From: ${moment.unix(traceResult.start).format() as String} - To: ${
+                moment.unix(traceResult.stop).format() as String
+            } 
+            - Order type: ${traceResult.orderType} - Amount: ${traceResult.traces.size}
+        """
+        )
+
+        when (js("traceOrderType") as String) {
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: You need to remove reliance on `js("traceOrderType")`. You have the real `traceOrderType` already in Kotlin.

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 6]
```diff
@@ -52,4 +115,428 @@ class TracesPage(
...
+            } else {
+                jq("#trace_table").append(rowHtml)
+            }
+        }
+
+        updateOccurredLabels()
+    }
+
+    private fun displayTraceStack(vararg traceStack: TraceSpanInfo) { //todo-chess-equality: [traceStack: List<TraceSpanInfo>]
+        js("portalLog('Displaying trace stack:');")
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: Same thing with relying on `js("portalLog(...))`. You have the real portalLog() method already in Kotlin.

In general, you should above `js()` as much as possible.

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 7]
```diff
@@ -52,4 +115,428 @@ class TracesPage(
...
+                }
+            } else {
+                jq("#trace_table").append(rowHtml)
+            }
+        }
+
+        updateOccurredLabels()
+    }
+
+    private fun displayTraceStack(vararg traceStack: TraceSpanInfo) { //todo-chess-equality: [traceStack: List<TraceSpanInfo>]
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: You can remove `todo-chess-equality`s when you no longer need them.

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 8]
```diff
@@ -52,4 +115,428 @@ class TracesPage(
             }
         }
     }
+
+    private fun setupUI() {
+        if (js("hideOverviewTab") as Boolean) {
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]: Any variables you access like this should be brought into the Kotlin page.
The same way `js("traceOrderType")` turned into:
```kotlin
class TracesPage(
    private val portalUuid: String,
    private val traceOrderType: TraceOrderType //<< this replaced js("traceOrderType")
)
```

File: portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt [hunk 9]
```diff
@@ -367,7 +357,7 @@ class TracesPage(
                             src = "../themes/default/assets/components/${component?.toUpperCase()}.png"
                         }
                         +spanInfo.operationName!!.replace("<", "&lt;").replace(">", "&gt;")
-                    } else if (span.hasChildStack!! || (!js("externalPortal") as Boolean && !span.artifactQualifiedName.isNullOrEmpty() && i > 0)) {
+                    } else if (span.hasChildStack!! || (externalPortal && !span.artifactQualifiedName.isNullOrEmpty() && i > 0)) {
```
Comment (BFergerson) [inline portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt]:
```suggestion
                    } else if (span.hasChildStack!! || (!externalPortal && !span.artifactQualifiedName.isNullOrEmpty() && i > 0)) {
```

**Comment by chess-equality** [comment]
@BFergerson What is the type of `log.data`? In `TraceSpanLogEntry` `log.data` is a string, but in js it is an object. Should a new object be created also for this?

```javascript
        if (log.data.event === 'error') {
            rowHtml += log.data.stack.replace(/(?:\r\n|\r|\n)/g, '<br>');
        } else {
            rowHtml += log.data;
        }
```

https://github.com/chess-equality/SourceMarker/blob/feature/issue-251/convert-js-to-kotlin/portal/backend/src/main/resources/webroot/js/views/traces_view.js#L153-L157

**Comment by BFergerson** [comment]
It's an object but it's not currently being handled correctly internally. See
https://github.com/sourceplusplus/SourceMarker/blob/49149ca3e4de7aa6439addd532bfda07b4fec248/monitor/skywalking/src/main/kotlin/com/sourceplusplus/monitor/skywalking/SkywalkingExtensions.kt#L67

You can create a new object or skip it altogether.

**Comment by chess-equality** [comment]
@BFergerson I moved below back to the method from `setupUI`, the tab doesn't change on click when the block is not on the method.

```javascript
jq("#span_info_panel").css("display", "none")
        jq("#latest_traces_header").addClass("active_sub_tab")
            .removeClass("inactive_tab")
        jq("#top_trace_table").css("display", "")
        jq("#trace_stack_table").css("visibility", "hidden")
        jq("#traces_span").css("display", "unset")
        jq("#trace_stack_span").css("display", "none")
        jq("#segment_id_span").css("display", "none")

        jq("#trace_stack_header").addClass("inactive_tab")
            .removeClass("active_sub_tab")
            .css("visibility", "hidden")

        jq("#span_info_header").addClass("inactive_tab")
            .removeClass("active_sub_tab")
            .css("visibility", "hidden")
```

https://github.com/chess-equality/SourceMarker/blob/feature/issue-251/convert-js-to-kotlin/portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt#L205-L220

**Comment by chess-equality** [comment]
@BFergerson Also if I enable below, I get a JavaScript error.

`// portalLog(JSON.parse(JSON.stringify(spanInfo)))`

https://github.com/chess-equality/SourceMarker/blob/feature/issue-251/convert-js-to-kotlin/portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt#L438

**Comment by chess-equality** [comment]
@BFergerson This is ready for review.

**Comment by BFergerson** [comment]
This is mostly good to go. As I was cleaning this up I stumbled upon the more correct way of working with KotlinJS that allows you to debug inside IntelliJ like normal code. Going to push those changes in this branch and merge and give you updated instructions on how to work with the portal module.

Till then feel free to start converting configuration. It's the easiest of the pages.

**Pull request merged**